File: II-155-2/BenfordsLaw.cs
using System;
using System.Globalization;
using System.Collections;
using System.Text;
using System.Text.RegularExpressions;

public class BenfordsLaw
{
	public int questionableDigit( int[] transactions, int threshold )
	{
		int[] freq = new int[10];
		for( int i = 0; i < transactions.Length; i++ )
		{
			int y = transactions[i];
			int x = y / 10;
			while( x > 0 )
			{
				y = x;
				x /= 10;
			}
			freq[y]++;
		}
		for( int i = 1; i < 10; i++ )
		{
			if ( (double) freq[i] / transactions.Length > (double) threshold * Math.Log10( 1 + 1.0 / i ) ||
				(double) freq[i] / transactions.Length < 1 / (double) threshold * Math.Log10( 1 + 1.0 / i ) )
			{
				return i;
			}
		}
		return -1;
	}

	NumberFormatInfo nfi = new CultureInfo( "en-US", false ).NumberFormat;

// BEGIN CUT HERE
	public static void Main(string[] args)
	{
		DateTime t1 = DateTime.Now;
		eq(1,(new BenfordsLaw()).questionableDigit(new int[] { 5236,7290,200,1907,3336,9182,17,4209,8746,7932,
  6375,909,2189,3977,2389,2500,1239,3448,6380,4812 }, 2),1);
		eq(2,(new BenfordsLaw()).questionableDigit(new int[] { 1,10,100,2,20,200,2000,3,30,300 }, 2),2);
		eq(3,(new BenfordsLaw()).questionableDigit(new int[] { 9,87,765,6543,54321,43219,321987,21987,1987,345,234,123 }, 2),-1);
		eq(4,(new BenfordsLaw()).questionableDigit(new int[] { 1,2,3,4,5,6,7,8,7,6,5,4,3,2,1 }, 8),9);
		eq(5,(new BenfordsLaw()).questionableDigit(new int[] { 987,234,1234,234873487,876,234562,17,
  7575734,5555,4210,678234,3999,8123 }, 3),8);
		DateTime t2 = DateTime.Now;
		TimeSpan t = t2 - t1;
		Console.WriteLine( "{0} sec", t.TotalMilliseconds / 1000 ); 
	}
	private static void eq(int n, int a, int b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, char a, char b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected '"+b+"', received '"+a+"'.");
	}
	private static void eq(int n, long a, long b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, bool a, bool b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, string a, string b)
	{
		if(a == b)
			Console.WriteLine("Case "+n+" passed.");
		else
			Console.WriteLine("Case "+n+" failed: expected \""+b+"\", received \""+a+"\".");
	}
	private static void eq(int n, int[] a, int[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, long[] a, long[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void eq(int n, string[] a, string[] b)
	{
		if(a.Length != b.Length)
		{
			Console.WriteLine("Case "+n+" failed: returned "+a.Length+" elements; expected "+b.Length+" elements.");
			return;
		}
		for(int i= 0; i < a.Length; i++)
			if(a[i] != b[i])
			{
				Console.WriteLine("Case "+n+" failed. Expected and returned array differ in position "+i);
				print(b);
				print(a);
				return;
			}
		Console.WriteLine("Case "+n+" passed.");
	}
	private static void print(int a)
	{
		Console.Write(a+" ");
	}
	private static void print(string s)
	{
		Console.Write("\""+s+"\" ");
	}
	private static void print(int[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(long[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write(rs[i]);
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void print(string[] rs)
	{
		if(rs == null) return;
		Console.Write('{');
		for(int i= 0; i < rs.Length; i++)
		{
			Console.Write("\""+rs[i]+"\"");
			if(i != rs.Length-1)
				Console.Write(", ");
		}
		Console.WriteLine('}');
	}
	private static void nl()
	{
		Console.WriteLine();
	}

// END CUT HERE
}

File: II-155-2/BenfordsLaw.cpp
#include <algorithm>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <iostream>
#include <iterator>
#include <map>
#include <numeric>
#include <set>
#include <sstream>
#include <stack>
#include <string>
#include <queue>
#include <vector>
using namespace std;
// BEGIN CUT HERE
#include <Windows.h>
#pragma comment( lib, "Winmm" )
// END CUT HERE
typedef vector<int> vi;
#define fora(i,v) for(int i = 0; i < (v).size(); i++)
struct BenfordsLaw
{
	int questionableDigit( vi transactions, int threshold )
	{
		vi freq(10, 0);
		fora(i, transactions)
		{
			int x;
			for( x = transactions[i]; x > 9; x /= 10 );
			++freq[x];
		}
		for( int result = 1; result <= 9; ++result )
			if ( static_cast<double>(freq[result]) / transactions.size() < log10( 1 + 1.0 / result ) / threshold ||
				 static_cast<double>(freq[result]) / transactions.size() > log10( 1 + 1.0 / result ) * threshold )
				return result;
		return -1;
	}
	// BEGIN CUT HERE
	private:
	template <typename T> string print_array( const vector<T>& v ) { ostringstream os; os << "{ "; for( typename vector<T>::const_iterator it = v.begin(); it != v.end(); ++it ) os << '\"' << *it << "\","; os << " }"; return os.str(); }
	template<typename T> void verify_case( int Case, const vector<T>& expected, const vector<T>& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: " << print_array( expected ) << endl; cerr << "\tReceived: " << print_array( received ) << endl; } }
	void verify_case( int Case, const long long& expected, const long long& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const double& expected, const double& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const int& expected, const int& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	void verify_case( int Case, const string& expected, const string& received ) { cerr << "Test Case #" << Case << "..."; if ( expected == received ) cerr << "PASSED. "; else { cerr << "FAILED." << endl; cerr << "\tExpected: \"" << expected << '\"' << endl; cerr << "\tReceived: \"" << received << '\"' << endl; } }
	public:
	void run_test( int Case = -1 ) { vector<vi> A0; vector<int> A1; vector<int> A2;
	{ /*Test 0*/int _transactions[] = 
	{ 5236,7290,200,1907,3336,9182,17,4209,8746,7932,
  6375,909,2189,3977,2389,2500,1239,3448,6380,4812 }
	; vi transactions(_transactions, _transactions + (sizeof(_transactions) / sizeof(_transactions[0]))); A0.push_back( transactions ); int threshold = 
	2
	; A1.push_back( threshold ); int result = 
	1
	; A2.push_back( result );  }
	{ /*Test 1*/int _transactions[] = 
	{ 1,10,100,2,20,200,2000,3,30,300 }
	; vi transactions(_transactions, _transactions + (sizeof(_transactions) / sizeof(_transactions[0]))); A0.push_back( transactions ); int threshold = 
	2
	; A1.push_back( threshold ); int result = 
	2
	; A2.push_back( result );  }
	{ /*Test 2*/int _transactions[] = 
	{ 9,87,765,6543,54321,43219,321987,21987,1987,345,234,123 }
	; vi transactions(_transactions, _transactions + (sizeof(_transactions) / sizeof(_transactions[0]))); A0.push_back( transactions ); int threshold = 
	2
	; A1.push_back( threshold ); int result = 
	-1
	; A2.push_back( result );  }
	{ /*Test 3*/int _transactions[] = 
	{ 1,2,3,4,5,6,7,8,7,6,5,4,3,2,1 }
	; vi transactions(_transactions, _transactions + (sizeof(_transactions) / sizeof(_transactions[0]))); A0.push_back( transactions ); int threshold = 
	8
	; A1.push_back( threshold ); int result = 
	9
	; A2.push_back( result );  }
	{ /*Test 4*/int _transactions[] = 
	{ 987,234,1234,234873487,876,234562,17,
  7575734,5555,4210,678234,3999,8123 }
	; vi transactions(_transactions, _transactions + (sizeof(_transactions) / sizeof(_transactions[0]))); A0.push_back( transactions ); int threshold = 
	3
	; A1.push_back( threshold ); int result = 
	8
	; A2.push_back( result );  }
	for( int i = 0; i < A0.size(); i++ ) if ( Case == -1 || i == Case ) { DWORD start, finish; start = timeGetTime(); verify_case( i, A2[i], questionableDigit( A0[i], A1[i]) ); finish = timeGetTime(); cout << "Elapsed " << finish - start << " msec" << endl; } }
// END CUT HERE
};
// BEGIN CUT HERE
int main()
{
	DWORD start, finish;
	start = timeGetTime();
	BenfordsLaw ___test; 
	___test.run_test(); 
	finish = timeGetTime();
	cout << endl << "Elapsed " << finish - start << " msec" << endl;
} 
// END CUT HERE
